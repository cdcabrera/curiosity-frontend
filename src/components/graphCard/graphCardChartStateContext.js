import React, { useContext } from 'react';
import { helpers } from '../../common/helpers';

/**
 * @memberof GraphCardChartState
 * @module GraphCardStateContext
 */

/**
 * GraphCardState context.
 *
 * @type {React.Context<{}>}
 */
const DEFAULT_CONTEXT = [{ chartDataSets: [] }, helpers.noop];

const GraphCardChartStateContext = React.createContext(DEFAULT_CONTEXT);

/**
 * Get an updated graph card context.
 *
 * @returns {React.Context<{}>}
 */
const useGraphCardStateChartContext = () => useContext(GraphCardChartStateContext);

const context = {
  GraphCardMetricTotalsContext: GraphCardChartStateContext,
  DEFAULT_CONTEXT,
  useGraphCardStateChartContext
};

export { context as default, context, GraphCardChartStateContext, DEFAULT_CONTEXT, useGraphCardStateChartContext };
