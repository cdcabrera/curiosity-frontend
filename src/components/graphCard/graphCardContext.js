import React, { useContext } from 'react';
import { reduxActions, storeHooks } from '../../redux';
import { useProduct, useProductGraphTallyQuery, useProductGraphConfig } from '../productView/productViewContext';
import { graphCardHelpers } from './graphCardHelpers';
import { helpers } from '../../common/helpers';

/**
 * Chart context.
 *
 * @type {React.Context<{}>}
 */
const DEFAULT_CONTEXT = [{ settings: { isStandalone: false, metrics: [], metric: undefined } }, helpers.noop];

const GraphCardContext = React.createContext(DEFAULT_CONTEXT);

/**
 * Get an updated graph card context.
 *
 * @returns {React.Context<{}>}
 */
const useGraphCardContext = () => useContext(GraphCardContext);

/**
 * Get grouped, or standalone, settings to be applied towards context.
 *
 * @param {Function} useProductGraphConfig
 * @returns {{ groupedContextSettings: {settings: object}, standaloneContextSettings: { id: string, settings: object }[] }}
 */
const useGardCardSettings = ({ useProductGraphConfig: useAliasProductGraphConfig = useProductGraphConfig } = {}) => {
  const { filters, settings } = useAliasProductGraphConfig();
  const { groupedFilters, standaloneFilters } = graphCardHelpers.generateChartSettings(filters);

  return {
    groupedContextSettings:
      (groupedFilters.length && {
        settings: {
          ...settings,
          isStandalone: false,
          metric: undefined,
          metrics: groupedFilters
        }
      }) ||
      undefined,
    standaloneContextSettings: standaloneFilters.map(metricFilter => ({
      id: metricFilter.id,
      settings: {
        padding: {
          bottom: 75,
          left: 75,
          right: 45,
          top: 45
        },
        ...settings,
        isStandalone: true,
        metric: metricFilter,
        metrics: [metricFilter]
      }
    }))
  };
};

/**
 * Use Redux RHSM Actions, getGraphTally.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProductGraphTallyQuery
 * @returns {Function}
 */
const useGetGraphTally = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery = useProductGraphTallyQuery
} = {}) => {
  const query = useAliasProductGraphTallyQuery();
  const dispatch = useAliasDispatch();

  return metrics => reduxActions.rhsm.getGraphTally(metrics, query)(dispatch);
};

/**
 * Get multiple metrics from store.
 *
 * @param {object} options
 * @param {Function} options.useGraphCardContext
 * @param {Function} options.useSelectors
 * @param {Function} options.useProduct
 * @returns {{data: object, pending: boolean, fulfilled: boolean, dataSets: Array, error: boolean}}
 */
const useMetricsSelector = ({
  useGraphCardContext: useAliasGraphCardContext = useGraphCardContext,
  useSelectors: useAliasSelectors = storeHooks.reactRedux.useSelectors,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const { settings = {} } = useAliasGraphCardContext();
  const { metrics = [] } = settings;
  const data = {};

  const metricResponses = useAliasSelectors(
    metrics.map(
      ({ id: metricId }) =>
        ({ graph }) =>
          graph.tally?.[`${productId}_${metricId}`]
    ),
    []
  );

  let isPending = false;
  let isFulfilled = false;
  let errorCount = 0;

  const dataSets = metricResponses.map((metric, index) => {
    const { pending, fulfilled, error, cancelled } = metric || {};
    const updatedPending = pending || cancelled || false;

    if (updatedPending) {
      isPending = true;
    }

    if (fulfilled) {
      isFulfilled = true;
    }

    if (error) {
      errorCount += 1;
    }

    const updatedMetric = {
      ...metrics[index],
      data: metric?.data?.data || [],
      meta: metric?.data?.meta || {}
    };
    data[metrics[index].id] = updatedMetric;

    return updatedMetric;
  });

  const response = {
    data,
    dataSets,
    error: false,
    fulfilled: false,
    pending: false
  };

  if (errorCount === dataSets.length) {
    response.error = true;
  } else if (isPending) {
    response.pending = true;
  } else if (isFulfilled) {
    response.fulfilled = true;
  }

  return response;
};

const context = {
  GraphCardContext,
  DEFAULT_CONTEXT,
  useGetGraphTally,
  useGraphCardContext,
  useGardCardSettings,
  useMetricsSelector
};

export {
  context as default,
  context,
  GraphCardContext,
  DEFAULT_CONTEXT,
  useGetGraphTally,
  useGraphCardContext,
  useGardCardSettings,
  useMetricsSelector
};
