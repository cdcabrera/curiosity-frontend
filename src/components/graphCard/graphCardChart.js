import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardActions, CardBody, CardHeader, CardTitle, Title } from '@patternfly/react-core';
import { useProduct, useProductGraphTallyQuery } from '../productView/productViewContext';
import { useGraphCardContext, useGetTally, useGetMetrics } from './graphCardContext';
import { graphCardHelpers } from './graphCardHelpers';
import { Chart } from '../chart/chart';
import { GraphCardChartLegend } from './graphCardChartLegend';
import { GraphCardChartTooltip } from './graphCardChartTooltip';
import { GraphCardChartTitleTooltip } from './graphCardChartTitleTooltip';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { MinHeight } from '../minHeight/minHeight';
import { Loader } from '../loader/loader';
import { translate } from '../i18n/i18n';

/**
 * A chart/graph card.
 *
 * @param {object} props
 * @param {Function} props.t
 * @param {Function} props.useGetGraphData
 * @param {Function} props.useGraphCardContext
 * @param {Function} props.useProduct
 * @param {Function} props.useProductGraphTallyQuery
 * @returns {Node}
 */
const GraphCardChart = ({
  t,
  useGetGraphData: useAliasGetGraphData,
  useGraphCardContext: useAliasGraphCardContext,
  useProduct: useAliasProduct,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery
}) => {
  const { productId } = useAliasProduct();
  const { settings = {} } = useAliasGraphCardContext();
  // const { actionDisplay, metric, metrics = [] } = settings;
  const { actionDisplay, metric } = settings;

  const { [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity } = useAliasProductGraphTallyQuery();
  const { pending, error, data = {}, dataSets = [] } = useAliasGetGraphData();
  /*
  const { pending, error, data = {}, dataSets = [] } = useAliasMetricsSelector();
  const getGraphTally = useAliasGetGraphTally();

  useShallowCompareEffect(() => {
    getGraphTally(metrics.map(({ id: metricId }) => ({ id: productId, metric: metricId })));
  }, [metrics, productId, getGraphTally]);
  */

  const standaloneMetricId = (metric?.id && `_${metric?.id}`) || '';
  let updatedActionDisplay = null;

  if (typeof actionDisplay === 'function') {
    updatedActionDisplay = actionDisplay({ data: { ...data } });
  }

  return (
    <Card className="curiosity-usage-graph">
      <MinHeight key="headerMinHeight">
        <CardHeader>
          <CardTitle>
            <Title headingLevel="h2" size="lg">
              {t(`curiosity-graph.cardHeading${standaloneMetricId}`, { context: productId })}
              <GraphCardChartTitleTooltip />
            </Title>
          </CardTitle>
          {updatedActionDisplay && (
            <CardActions className={(error && 'blur') || ''}>{updatedActionDisplay}</CardActions>
          )}
        </CardHeader>
      </MinHeight>
      <MinHeight key="bodyMinHeight">
        <CardBody>
          <div className={(error && 'blur') || (pending && 'fadein') || ''}>
            {pending && <Loader variant="graph" />}
            {!pending && (
              <Chart
                {...graphCardHelpers.generateExtendedChartSettings({ settings, granularity })}
                dataSets={dataSets}
                chartLegend={({ chart, datum }) => <GraphCardChartLegend chart={chart} datum={datum} />}
                chartTooltip={({ datum }) => <GraphCardChartTooltip datum={datum} />}
              />
            )}
          </div>
        </CardBody>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{useGraphCardContext: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     t: Function, useGetGraphData: Function}}
 */
GraphCardChart.propTypes = {
  t: PropTypes.func,
  useGetGraphData: PropTypes.func,
  // useGetGraphTally: PropTypes.func,
  useGraphCardContext: PropTypes.func,
  // useMetricsSelector: PropTypes.func,
  useProduct: PropTypes.func,
  useProductGraphTallyQuery: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useGraphCardContext: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     t: translate, useGetGraphData: Function}}
 */
GraphCardChart.defaultProps = {
  t: translate,
  useGetGraphData: useGetMetrics,
  // useGetGraphTally,
  useGraphCardContext,
  useProduct,
  useProductGraphTallyQuery
};

export { GraphCardChart as default, GraphCardChart };
