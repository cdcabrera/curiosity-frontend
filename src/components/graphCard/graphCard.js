import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardTitle, CardHeader, CardActions, CardBody, Title } from '@patternfly/react-core';
import { useDeepCompareEffect, useShallowCompareEffect } from 'react-use';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { helpers } from '../../common';
import { RHSM_API_QUERY_GRANULARITY_TYPES as GRANULARITY_TYPES, RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { Loader } from '../loader/loader';
import { MinHeight } from '../minHeight/minHeight';
import { translate } from '../i18n/i18n';
import { GraphCardChart } from './graphCardChart';

/**
 * A chart/graph card.
 *
 * @param {object} props
 * @param {string} props.cardTitle
 * @param {Node} props.children
 * @param {boolean} props.error
 * @param {Array} props.filterGraphData
 * @param {boolean} props.fulfilled
 * @param {Function} props.getGraphReportsCapacity
 * @param {object} props.graphData
 * @param {boolean} props.isDisabled
 * @param {boolean} props.pending
 * @param {string} props.productId
 * @param {string} props.productLabel
 * @param {object} props.query
 * @param {string} props.viewId
 * @returns {Node}
 */
const GraphCard = ({
  cardTitle,
  children,
  error,
  filterGraphData,
  getGraphReportsCapacity,
  graphData,
  isDisabled,
  pending,
  productId,
  productLabel,
  query,
  viewId
}) => {
  // const { pending, fulfilled, error } = useSelector(({ graph }) => graph?.reportCapacity?.[productId], {});
  const [updatedGraphData, setUpdatedGraphData] = React.useState(graphData);
  const {
    [RHSM_API_QUERY_TYPES.START_DATE]: startDate,
    [RHSM_API_QUERY_TYPES.END_DATE]: endDate,
    [RHSM_API_QUERY_TYPES.GRANULARITY]: granularity
  } = query;

  useShallowCompareEffect(() => {
    if (!isDisabled && startDate && endDate && granularity && productId) {
      getGraphReportsCapacity(productId, query);
    }
  }, [isDisabled, startDate, endDate, granularity, productId, getGraphReportsCapacity, query]);

  useDeepCompareEffect(() => {
    setUpdatedGraphData(graphData);
  }, [graphData]);

  if (isDisabled) {
    return null;
  }

  return (
    <Card className="curiosity-usage-graph">
      <MinHeight key="headerMinHeight">
        <CardHeader>
          <CardTitle>
            <Title headingLevel="h2" size="lg">
              {cardTitle}
            </Title>
          </CardTitle>
          <CardActions className={(error && 'blur') || ''}>{children}</CardActions>
        </CardHeader>
      </MinHeight>
      <MinHeight key="bodyMinHeight">
        <CardBody>
          <div className={(error && 'blur') || (pending && 'fadein') || ''}>
            {pending && <Loader variant="graph" />}
            {!pending && (
              <GraphCardChart
                filterGraphData={filterGraphData}
                granularity={granularity}
                graphData={updatedGraphData}
                productLabel={productLabel}
                viewId={viewId}
              />
            )}
          </div>
        </CardBody>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{productLabel: string, productId: string, pending: boolean, error: boolean, query: object,
 *     cardTitle: string, filterGraphData: Array, getGraphReportsCapacity: Function,
 *     viewId: string, t: Function, children: Node, graphData: object, isDisabled: boolean}}
 */
GraphCard.propTypes = {
  cardTitle: PropTypes.string,
  children: PropTypes.node,
  error: PropTypes.bool,
  filterGraphData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      fill: PropTypes.string,
      stroke: PropTypes.string
    })
  ),
  // fulfilled: PropTypes.bool,
  getGraphReportsCapacity: PropTypes.func,
  graphData: PropTypes.object,
  query: PropTypes.shape({
    [RHSM_API_QUERY_TYPES.GRANULARITY]: PropTypes.oneOf([...Object.values(GRANULARITY_TYPES)]).isRequired,
    [RHSM_API_QUERY_TYPES.START_DATE]: PropTypes.string.isRequired,
    [RHSM_API_QUERY_TYPES.END_DATE]: PropTypes.string.isRequired
  }).isRequired,
  isDisabled: PropTypes.bool,
  pending: PropTypes.bool,
  productId: PropTypes.string.isRequired,
  productLabel: PropTypes.string,
  t: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{getGraphReportsCapacity: Function, productLabel: string, viewId: string, t: translate,
 *     children: Node, pending: boolean, graphData: object, isDisabled: boolean, error: boolean,
 *     cardTitle: string, filterGraphData: Array}}
 */
GraphCard.defaultProps = {
  cardTitle: null,
  children: null,
  error: false,
  filterGraphData: [],
  // fulfilled: false,
  getGraphReportsCapacity: helpers.noop,
  graphData: {},
  isDisabled: helpers.UI_DISABLED_GRAPH,
  pending: false,
  productLabel: '',
  t: translate,
  viewId: 'graphCard'
};

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getGraphReportsCapacity: (id, query) => dispatch(reduxActions.rhsm.getGraphReportsCapacity(id, query))
});

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.graphCard.makeGraphCard();

const ConnectedGraphCard = connect(makeMapStateToProps, mapDispatchToProps)(GraphCard);

export { ConnectedGraphCard as default, ConnectedGraphCard, GraphCard };

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
// const makeMapStateToProps = reduxSelectors.graphCard.makeGraphCard();
/*
const mapStateToProps = (state, props) => {
  const { graphTallyQuery: query } = apiQueries.parseRhsmQuery(
    {
      ...props.query,
      ...state.view?.query?.[props.productId],
      ...state.view?.query?.[props.viewId]
    },
    {
      graphTallyQuery: { ...state.view?.graphTallyQuery?.[props.viewId] }
    }
  );

  return { ...state.graph?.reportCapacity?.[props.productId], query };
};
*/
/*
const makeMapStateToProps = () => (state, props) => {
  const { graphTallyQuery: query } = apiQueries.parseRhsmQuery(
    {
      ...props.query,
      ...state.view?.query?.[props.productId],
      ...state.view?.query?.[props.viewId]
    },
    {
      graphTallyQuery: { ...state.view?.graphTallyQuery?.[props.viewId] }
    }
  );
  const { pending, fulfilled, error } = state.graph?.reportCapacity?.[props.productId] || {};

  return { pending, fulfilled, error, query };
};
*/

/*
const mapStateToProps = (state, props) => {
  const { graphTallyQuery: query } = apiQueries.parseRhsmQuery(
    {
      ...props.query,
      ...state.view?.query?.[props.productId],
      ...state.view?.query?.[props.viewId]
    },
    {
      graphTallyQuery: { ...state.view?.graphTallyQuery?.[props.viewId] }
    }
  );
  const { pending, fulfilled, error } = state.graph?.reportCapacity?.[props.productId];

  return { pending, fulfilled, error, query };
};
*/
