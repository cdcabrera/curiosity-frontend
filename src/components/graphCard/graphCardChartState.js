import React, { useMemo, useState } from 'react';
import { GraphCardMetricTotals } from './graphCardMetricTotals';
import { GraphCardChart } from './graphCardChart';
import { GraphCardChartStateContext } from './graphCardChartStateContext';

/**
 * Wrapper for providing up-to-date internal state/context for graphCard components.
 *
 * @see Charts
 * @memberof Components
 * @module GraphCardState
 * @property {module} GraphCardStateContext
 */

/**
 * Set up graph cards. Expand filters with base graph settings.
 *
 * @returns {React.ReactNode}
 */
const GraphCardChartState = () => {
  const [context, setContext] = useState();
  const contextValue = useMemo(() => [context, setContext], [context, setContext]);

  return (
    <GraphCardChartStateContext.Provider value={contextValue}>
      <GraphCardMetricTotals>
        <GraphCardChart />
      </GraphCardMetricTotals>
    </GraphCardChartStateContext.Provider>
  );
};

/**
 * Prop types.
 *
 * @type {{}}
 */
GraphCardChartState.propTypes = {};

/**
 * Default props.
 *
 * @type {{}}
 */
GraphCardChartState.defaultProps = {};

export { GraphCardChartState as default, GraphCardChartState };
