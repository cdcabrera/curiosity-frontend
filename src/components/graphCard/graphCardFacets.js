import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardTitle, CardHeader, CardActions, CardBody, Title } from '@patternfly/react-core';
import { useShallowCompareEffect } from 'react-use';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { useProduct, useProductGraphConfig, useProductGraphTallyQuery } from '../productView/productViewContext';
import { helpers } from '../../common';
import { RHSM_API_QUERY_TYPES } from '../../types/rhsmApiTypes';
import { Loader } from '../loader/loader';
import { MinHeight } from '../minHeight/minHeight';
import { GraphCardChart } from './graphCardChart';

/**
 * A chart/graph card.
 * This card makes use of the new Tally API response format. Eventually this component will replace "GraphCard".
 *
 * @param {object} props
 * @param {Node} props.cardTitle
 * @param {Node} props.children
 * @param {boolean} props.error
 * @param {Function} props.getGraphTally
 * @param {object} props.graphData
 * @param {object} props.meta
 * @param {boolean} props.isDisabled
 * @param {boolean} props.pending
 * @param {Function} props.useProduct
 * @param {Function} props.useProductGraphConfig
 * @param {Function} props.useProductGraphTallyQuery
 * @returns {Node}
 */
const GraphCardFacets = ({
  cardTitle,
  children,
  error,
  getGraphTally,
  graphData,
  meta,
  isDisabled,
  pending,
  useProduct: useAliasProduct,
  useProductGraphConfig: useAliasProductGraphConfig,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery
}) => {
  const { productId } = useAliasProduct();
  const { filters, settings } = useAliasProductGraphConfig();
  const query = useAliasProductGraphTallyQuery();

  useShallowCompareEffect(() => {
    const {
      [RHSM_API_QUERY_TYPES.START_DATE]: startDate,
      [RHSM_API_QUERY_TYPES.END_DATE]: endDate,
      [RHSM_API_QUERY_TYPES.GRANULARITY]: granularity
    } = query;

    if (!isDisabled && filters.length && granularity && startDate && endDate && productId) {
      // need to reconsider where this is called... "regular card" which would group all the metrics vs "facet card" which would only need a single metric
      getGraphTally(
        filters.map(({ id: metric }) => ({ id: productId, metric })),
        query
      );
    }
  }, [filters, getGraphTally, isDisabled, productId, query]);

  if (isDisabled) {
    return null;
  }

  const standaloneFacets = filters.filter(({ isStandalone }) => isStandalone === true);
  const regularFacets = filters.filter(({ isStandalone }) => isStandalone !== true);

  const regularCard = () => {
    let actionDisplay = null;

    if (typeof settings?.actionDisplay === 'function') {
      actionDisplay = settings.actionDisplay({ data: { ...graphData }, meta: { ...meta } });
    }

    return (
      <Card className="curiosity-usage-facet-graph">
        <MinHeight key="headerMinHeight">
          <CardHeader>
            <CardTitle>
              <Title headingLevel="h2" size="lg">
                {cardTitle}
              </Title>
            </CardTitle>
            <CardActions className={(error && 'blur') || ''}>
              <React.Fragment key="actionDisplay">{actionDisplay}</React.Fragment>
              {children}
            </CardActions>
          </CardHeader>
        </MinHeight>
        <MinHeight key="bodyMinHeight">
          <CardBody>
            <div className={(error && 'blur') || (pending && 'fadein') || ''}>
              {pending && <Loader variant="graph" />}
              {!pending && <GraphCardChart graphData={regularFacets} />}
            </div>
          </CardBody>
        </MinHeight>
      </Card>
    );
  };

  return <React.Fragment>{regularFacets.length && regularCard()}</React.Fragment>
};

/**
 * Prop types.
 *
 * @type {{getGraphReportsCapacity: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     useProductGraphConfig: Function, children: Node, meta: object, pending: boolean, graphData: object,
 *     isDisabled: boolean, error: boolean, cardTitle: Node}}
 */
GraphCardFacets.propTypes = {
  cardTitle: PropTypes.node,
  children: PropTypes.node,
  error: PropTypes.bool,
  getGraphTally: PropTypes.func,
  graphData: PropTypes.object,
  isDisabled: PropTypes.bool,
  meta: PropTypes.object,
  pending: PropTypes.bool,
  useProduct: PropTypes.func,
  useProductGraphConfig: PropTypes.func,
  useProductGraphTallyQuery: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{getGraphReportsCapacity: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     useProductGraphConfig: Function, children: Node, meta: object, pending: boolean, graphData: object,
 *     isDisabled: boolean, error: boolean, cardTitle: Node}}
 */
GraphCardFacets.defaultProps = {
  cardTitle: null,
  children: null,
  error: false,
  getGraphTally: helpers.noop,
  graphData: {},
  isDisabled: helpers.UI_DISABLED_GRAPH,
  meta: {},
  pending: false,
  useProduct,
  useProductGraphConfig,
  useProductGraphTallyQuery
};

/**
 * Apply actions to props.
 *
 * @param {Function} dispatch
 * @returns {object}
 */
const mapDispatchToProps = dispatch => ({
  getGraphTally: (id, query) => dispatch(reduxActions.rhsm.getGraphTally(id, query))
});

/**
 * Create a selector from applied state, props.
 *
 * @type {Function}
 */
const makeMapStateToProps = reduxSelectors.graph.makeGraph();

const ConnectedGraphCardColumns = connect(makeMapStateToProps, mapDispatchToProps)(GraphCardFacets);

export { ConnectedGraphCardColumns as default, ConnectedGraphCardColumns, GraphCardFacets };
