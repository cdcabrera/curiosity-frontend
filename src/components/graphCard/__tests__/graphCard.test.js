import React from 'react';
import { GraphCard } from '../graphCard';

describe('GraphCard Component', () => {
  it('should setup basic settings', () => {
    const props = {
      useParseFiltersSettings: () => ({
        filtersSettings: [
          {
            settings: {
              loremIpsum: true,
              metric: undefined,
              metrics: [
                {
                  chartType: 'area',
                  id: 'Sockets',
                  isCapacity: false,
                  isStacked: true,
                  loremIpsum: true,
                  isThreshold: false,
                  isToolbarFilter: false,
                  metric: 'Sockets',
                  strokeWidth: 2
                }
              ],
              padding: {
                bottom: 75,
                left: 75,
                right: 45,
                top: 45
              }
            }
          }
        ]
      })
    };
    const componentStandalone = renderComponent(<GraphCard {...props} />);
    expect(componentStandalone).toMatchSnapshot('settings, standalone');
    componentStandalone.unmount();

    props.useParseFiltersSettings = () => ({
      filtersSettings: [
        {
          settings: {
            loremIpsum: false,
            isMetricDisplay: true,
            cards: [
              {
                header: 'lorem',
                body: 'ipsum',
                footer: 'dolor sit'
              }
            ],
            metric: undefined,
            metrics: [
              {
                chartType: 'area',
                id: 'Core-seconds',
                isCapacity: false,
                isStacked: true,
                loremIpsum: false,
                isThreshold: false,
                isToolbarFilter: false,
                metric: 'Core-seconds',
                strokeWidth: 2
              }
            ]
          }
        }
      ]
    });

    const componentGrouped = renderComponent(<GraphCard {...props} />);
    expect(componentGrouped).toMatchSnapshot('settings, grouped');
  });

  it('should allow being disabled', () => {
    const props = {
      isDisabled: true
    };
    const component = renderComponent(<GraphCard {...props} />);
    expect(component).toMatchSnapshot('disabled');
  });
});
