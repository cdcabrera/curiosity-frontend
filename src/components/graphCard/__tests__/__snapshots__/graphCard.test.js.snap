// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCard Component should allow being disabled: disabled 1`] = `
<graphcard
  props="{
  "isDisabled": true,
  "children": []
}"
/>
`;

exports[`GraphCard Component should setup basic settings: settings, grouped 1`] = `
<graphcard
  props="{
  "isDisabled": false,
  "children": []
}"
>
  <div
    class="curiosity-usage-graph__totals"
    data-test="graphMetricTotals-undefined"
    data-test-data="{}"
  >
    <div>
      <div
        class="curiosity-usage-graph__totals-column"
      >
        <article
          class="pf-c-card pf-m-plain curiosity-usage-graph__totals-column-card "
          data-ouia-component-id="OUIA-Generated-Card-2"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe="true"
          data-test="graphMetricTotalsCard-0"
          id=""
        >
          <div
            class="pf-c-card__header"
          >
            <div
              class="pf-c-card__title"
            >
              <h2
                class="pf-c-title pf-m-md"
                data-ouia-component-id="OUIA-Generated-Title-2"
                data-ouia-component-type="PF4/Title"
                data-ouia-safe="true"
              />
            </div>
          </div>
          <div
            class="pf-c-card__body"
          >
            <div />
          </div>
          <div
            class="pf-c-card__footer"
          >
            <div />
          </div>
        </article>
      </div>
    </div>
    <div>
      <div
        class="curiosity-usage-graph__totals-graph-column"
      >
        <article
          class="pf-c-card pf-m-plain curiosity-usage-graph"
          data-ouia-component-id="OUIA-Generated-Card-3"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe="true"
          id=""
        >
          <div
            class="pf-c-card__header"
          >
            <div
              class="pf-c-card__title"
            >
              <h2
                class="pf-c-title pf-m-lg"
                data-ouia-component-id="OUIA-Generated-Title-3"
                data-ouia-component-type="PF4/Title"
                data-ouia-safe="true"
              >
                t(curiosity-graph.cardHeading, {})
              </h2>
            </div>
          </div>
          <div
            class="curiosity-minheight"
          >
            <div>
              <div
                class="pf-c-card__body"
              >
                <div
                  class=""
                >
                  <div
                    class="curiosity-chartarea uxui-curiosity__modal uxui-curiosity__modal--loading"
                    data-test="curiosity-chartarea"
                  />
                </div>
              </div>
            </div>
          </div>
        </article>
      </div>
    </div>
  </div>
</graphcard>
`;

exports[`GraphCard Component should setup basic settings: settings, standalone 1`] = `
<graphcard
  props="{
  "isDisabled": false,
  "children": []
}"
>
  <div
    class="curiosity-usage-graph__standalone"
    data-test="graphStandalone"
  >
    <article
      class="pf-c-card pf-m-plain curiosity-usage-graph"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF4/Card"
      data-ouia-safe="true"
      id=""
    >
      <div
        class="pf-c-card__header"
      >
        <div
          class="pf-c-card__title"
        >
          <h2
            class="pf-c-title pf-m-lg"
            data-ouia-component-id="OUIA-Generated-Title-1"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
          >
            t(curiosity-graph.cardHeading, {})
          </h2>
        </div>
      </div>
      <div
        class="curiosity-minheight"
      >
        <div>
          <div
            class="pf-c-card__body"
          >
            <div
              class=""
            >
              <div
                class="curiosity-chartarea uxui-curiosity__modal uxui-curiosity__modal--loading"
                data-test="curiosity-chartarea"
              />
            </div>
          </div>
        </div>
      </div>
    </article>
  </div>
</graphcard>
`;
